api/src/main/java/org/openmrs/CohortMembership.java

```diff
@@ -101,4 +120,35 @@ public void setEndDate(Date endDate) {
 	public int compareTo(CohortMembership o) {
 		return this.getPatient().getPatientId() - o.getPatient().getPatientId();
 	}
+
+	/**
+	 * @should compare two objects for equality
+	 */
+	@Override
+	public boolean equals(Object obj) {
+		if (obj instanceof CohortMembership) {
+			CohortMembership o = (CohortMembership) obj;
+			if (this.getUuid().equals(o.getUuid())) {
+				return super.equals(obj);
+			} else {
+				return Objects.equals(this.getPatient().getPatientId(), o.getPatient().getPatientId())
+						&& Objects.equals(this.getCohort().getCohortId(), o.getCohort().getCohortId())
+						&& this.getStartDate().equals(o.getStartDate())
+						&& ((this.getEndDate() != null && o.getEndDate() != null
+						&& this.getEndDate().equals(o.getEndDate()))
+						|| (this.getEndDate() == null && o.getEndDate() == null));
+			}
+		}
+		return false;
+	}
+	
+	@Override
+	public int hashCode() {
+		int result = 0;
+		result = 71 * result + (this.getPatient().getPatientId() != null
+										? this.getPatient().getPatientId().hashCode() : 0);
+		result = 71 * result + (this.getCohort().getCohortId() != null
```

Comment:
Implementation of hashcode() needs to align with equals, in the equals method you do delegate to super and you would need to do the same here but this might complicate the logic. On the other hand, I think you could move this logic to the compareTo() method and not have to implement equals() and hashcode() since Set uses compareTo and not equals